spec/index.bs

```diff
@@ -1518,20 +1500,184 @@ enum SupportedType {
+
+            5.  If |policy|'s [=policy/disposition=] is "`enforce`", then set |result| to
+                "`Blocked`".
+
+    3.  If |result| is "`Blocked`", throw an `EvalError` exception.
+
+10. <ins>Return |sourceString|.</ins>
+
+Note: returning |sourceString| means that the string that gets
+compiled is that returned by any [=default policy=] in the course of
+executing [$Get Trusted Type compliant string$].
+
+Issue: This depends on a solution to
+<a href="https://github.com/WICG/trusted-types/issues/144">issue #144</a> like <a
+href="https://mikesamuel.github.io/proposal-hostensurecancompilestrings-passthru/">TC39 HostBeforeCompile</a>
+
+Issue: In some cases, the violation "`'report-sample'`" contain the result of
+applying the default policy to a string argument which differs.
+Specifically when, there is a [=default policy=], |isExempt| is false,
+and |source| there is a CSP policy for either the |callerRealm|
+or |callerRealm| that disallows "`'unsafe-eval'"`.
+Is this a feature or a bug?
+
+Note: The previous algorithm reports violations via both report-uris where
+callerRealm != calleeRealm.  If [$Get Trusted Type compliant string$] reports an
+error, it only reports it via its |calleeRealm|'s report-uri.
+
+
+### <dfn abstract-op>IsSourceExempt</dfn> Algorithm ### {#is-source-exempt-algorithm}
+
+The IsSourceExempt algorithm takes a list of
+[[CSP3#content-security-policy-object|policies]] (|cspPolicies|),
+and a {{TrustedTypeConfiguration}} (|ttConfig|) and executes the
+following steps:
+
+1.  Let |isTTStrict| = `true`.
+1.  If |ttConfig|'s {{TrustedTypeConfiguration/unknownPolicyName}} is
+    {{UnknownPolicyNameDisposition|"allow"}} then
+    set |isTTStrict| to `false`.
+1.  Otherwise if |ttConfig|'s {{TrustedTypeConfiguration/allowedNames}} contains `'*'` then
+    set |isTTStrict| to `false`.
+1.  Let |isTTEnforced| = `false`.
+1.  If |ttConfig|'s {{TrustedTypeConfiguration/domSinks}} is
+    {{StringAtSinkDisposition|"reject"}} then
+    set |isTTEnforced| to `true`.
+1.  Let |isCSPEnforced| = `false`.
+1.  For each |policy| in |cspPolicies|:
+    1.  Let |source-list| be `null`.
+    1.  If |policy| contains a [[CSP3#directive|directive]]
+        whose [[CSP3#directive-name|name]] is "`script-src`", then set
+        |source-list| to that [[CSP3#directive|directive]]'s
+        [[CSP3#directive-value|value]].
+
+        Otherwise if |policy| contains a [[CSP3#directive|directive]]
+        whose [[CSP3#directive-name|name]] is "`default-src`", then set
+        |source-list| to that [[CSP3#directive|directive]]'s
+        [[CSP3#directive-value|value]].
+    1.  If |source-list| is not `null`, and does not contain a
+        [[CSP3#source-expression|source expression]] which as an
+        <a>ASCII case-insensitive</a> match for
+        the string "<a grammar>'unsafe-eval'</a>", then
+        set |isCSPEnforced| to `true`.
+1.  If |isTTStrict| is `true` and |isCSPEnforced| is equal to
```

Comment:
Why do you calculate & check for `isCSPEnforced` here? Can't `eval(TrustedScript)` be exempt from CSP restrictions if `isTTStrict` and `isTTEnforced` under script-restricting CSP, and always otherwise? 

IIUC, that would happen if you don't do any CSP-related calculations in this algorithm. `EnsureCSPDoesNotBlockStringCompilation` will enforce any CSP restrictions if this algorithm returned false, which may simply do nothing if there are no CSPs set.

What usecase does the condition in this step cover?

Reply:
I'm looking for consistency between enforcement, because I want testing with TT & CSP in report-only mode to only report violations during normal use that need to be fixed to avoid violation reports with enforcement on during normal use.

So I worry that if I don't check for consistency between TT and CSP that an engineer migrating would see CSP report-only violation reports that would be suppressed by the is-exempt check were TT and CSP both in enforce mode.

If that sounds like a good goal, maybe this deserves a note.

Reply:
So this assumes you're swapping the CSP enforcement together with TT enforcement? For example, if you have enforcing TT and report-only CSP, this will generate violations that you have no way of fixing, which would disappear when CSP is enforced. 

In general, during rollouts it would be rare to have the disposition of both CSP and TT to be changed simultaneously.